src/gtest/test_sidechain.cpp

```diff
@@ -1020,6 +1037,107 @@ TEST_F(SidechainTestSuite, CSidechainFromMempoolRetrievesUnconfirmedInformation)
     EXPECT_TRUE(retrievedInfo.mImmatureAmounts.at(-1) == fwdAmount);
 }
 
+TEST_F(SidechainTestSuite, CSidechainBlockUndoVersioning) {
+
+    static const std::string autofileName = "/tmp/test_block_undo_versioning.txt";
+    CAutoFile fileout(fopen(autofileName.c_str(), "w+") , SER_DISK, CLIENT_VERSION);
+    EXPECT_TRUE(fileout.Get() != NULL);
+
+    // write an old version undo block to the file
+    //----------------------------------------------
+    CBlockUndo_OldVersion buov;
+    buov.vtxundo.reserve(1);
+    buov.vtxundo.push_back(CTxUndo());
+
+    fileout << buov;;
+
+    uint256 h_buov;
+    {
+        CHashWriter hasher(SER_GETHASH, PROTOCOL_VERSION);
+        hasher << buov;
+        h_buov = hasher.GetHash();
+    }
+    fileout << h_buov;
+
+    fseek(fileout.Get(), 0, SEEK_END);
+    unsigned long len = (unsigned long)ftell(fileout.Get());
+
+    unsigned long buov_sz = buov.GetSerializeSize(SER_DISK, CLIENT_VERSION);
+    EXPECT_TRUE(len == buov_sz + sizeof(uint256));
+    
+    // write a new version undo block to the same file
+    //-----------------------------------------------
+    CBlockUndo buon;
+    buon.vtxundo.reserve(1);
+    buon.vtxundo.push_back(CTxUndo());
+
+    fileout << buon;;
+
+    uint256 h_buon;
+    {
+        CHashWriter hasher(SER_GETHASH, PROTOCOL_VERSION);
+        hasher << buon;
+        h_buon = hasher.GetHash();
+    }
+    fileout << h_buon;
+
+    fseek(fileout.Get(), 0, SEEK_END);
+    unsigned long len2 = (unsigned long)ftell(fileout.Get());
+
+    unsigned long buon_sz = buon.GetSerializeSize(SER_DISK, CLIENT_VERSION);
+    EXPECT_TRUE(len2 == len + buon_sz + sizeof(uint256));
+    
+    EXPECT_TRUE(buov_sz != buon_sz);
+
+    fileout.fclose();
+
+    // read both blocks and tell their version
+    //-----------------------------------------------
+    CAutoFile filein(fopen(autofileName.c_str(), "r+") , SER_DISK, CLIENT_VERSION);
+    EXPECT_TRUE(filein.Get() != NULL);
+
+    bool good_read = true;
+    CBlockUndo b1, b2;
+    uint256 h1, h2;
+    try {
+        filein >> b1;
+        filein >> h1;
+        filein >> b2;
+        filein >> h2;
+    }
+    catch (const std::exception& e) {
+        good_read = false;
+    }
+
+    EXPECT_TRUE(good_read == true);
+
+    EXPECT_TRUE(b1.IsNewVersion() == false);
+    EXPECT_TRUE(h1 == h_buov);
+
+    EXPECT_TRUE(b2.IsNewVersion() == true);
+    EXPECT_TRUE(h2 == h_buon);
+
+    // test Rewind method by reading the last hash again
+    filein.Rewind(sizeof(uint256));
+    uint256 h3;
+    EXPECT_TRUE(h2 != h3);
+    filein >> h3;
+    EXPECT_TRUE(h2 == h3);
+
+    // rewind by an impossible offset
+    try
+    {
+        filein.Rewind(len2+1);
+    }
+    catch(...)
+    {
+        good_read = false;
+    }
+    EXPECT_TRUE(good_read == false);
+
+    filein.fclose();
+}
+
```

Comment:
If we go for the `::AddEntriesInVector ` alternative, I'd restructure this UTs to prove the following points:

1. The new serialization/deserialization pair can successfully write and read in memory CBlocksUndo with and without sidechain related attributes
2. Given an explicit string represented the serialized version of the pre-sidechain CBlockUndo, the new serialization ops handle it allright.
Check out /src/test/coins_tests.cpp for a reference

Reply:
I will add that test to existing one